### PR TITLE
Dev/3.3.1

### PR DESCRIPTION
**Pull request by alexjhawk**

### CODE REVIEW

**Comment by alexjhawk** [comment]
@TomKimsey CHANGELOG.md updated to reflect a format similar to the requested changes on the Cumulocity connector.

File: src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java [hunk 1]
```diff
@@ -121,114 +125,193 @@ private static void runGrabData() {
...
+                            (int) queueBehindMillis,
+                            DAYS_STRING,
+                            HOURS_STRING,
+                            MINUTES_STRING,
+                            SECONDS_STRING);
+                    Logger.LOG_WARN(
+                        "The historical data queue is running behind by " + timeBehindString + ".");
+                  } else {
+                    // Queue is not past running behind threshold, reset to 0 for updating debug tag
+                    queueBehindMillis = 0;
```
Comment (TomKimsey) [inline src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java:192]: 0 could be TWConnectorConsts.QUEUE_DATA_POLL_CURRENT_MILLIS or something similar.

File: src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java [hunk 2]
```diff
@@ -121,114 +125,193 @@ private static void runGrabData() {
...
+                  }
+
+                  // Update available memory variable
+                  availableMemoryBytes = Runtime.getRuntime().freeMemory();
+
+                  // Enable queue double data rate if running behind and memory is available
+                  if (!queueDoubleDataRateEnabled
+                      && queueBehindMillis > 0
```
Comment (TomKimsey) [inline src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java:200]: TWConnectorConsts.QUEUE_DATA_POLL_CURRENT_MILLIS

File: src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java [hunk 3]
```diff
@@ -121,114 +125,193 @@ private static void runGrabData() {
...
+                    HistoricalDataQueueManager.setQueueFifoTimeSpanMins(
+                        doubleQueueDataPollSizeMins);
+                    Logger.LOG_SERIOUS(
+                        "The size of data polled on each interval has been doubled to "
+                            + doubleQueueDataPollSizeMins
+                            + " minutes while the queue is running behind!");
+                    queueDoubleDataRateEnabled = true;
+                  }
+                  // Disable queue double data rate if no longer running behind
+                  else if (queueDoubleDataRateEnabled && queueBehindMillis == 0) {
```
Comment (TomKimsey) [inline src/main/java/com/hms_networks/americas/sc/thingworx/TWConnectorMain.java:214]: TWConnectorConsts.QUEUE_DATA_POLL_CURRENT_MILLIS

**Review by TomKimsey — APPROVED** [review]
Consider making that 0 that shows up a constant. Approved as its not in the scope of this PR.

**Review by oliver-walker-hms — APPROVED** [review]

**Comment by alexjhawk** [comment]
@TomKimsey Added https://github.com/hms-networks/flexy-thingworx-connector/issues/69 to address your suggestion in a future PR.

**Comment by alexjhawk** [comment]
fixes #67 closes #67

**Pull request merged**